contracts/modules/SwapHub.sol

```diff
@@ -0,0 +1,182 @@
+    /// @param params struct defining the requested trade
+    function swap(uint subAccountIdIn, uint subAccountIdOut, address swapHandler, ISwapHandler.SwapParams memory params) external nonReentrant {
+        SwapCache memory cache = initSwap(subAccountIdIn, subAccountIdOut, params);
+
+        emit RequestSwapHub(
+            cache.accountIn,
+            cache.accountOut,
+            params.underlyingIn,
+            params.underlyingOut,
+            params.amountIn,
+            params.amountOut,
+            params.mode,
+            swapHandler
+        );
+
+        uint amountOut = swapInternal(cache, swapHandler, params);
+
+        // Process deposit
+        uint amountOutDecoded = decodeExternalAmount(cache.assetCacheOut, amountOut);
+        uint amountOutInternal = underlyingAmountToBalance(cache.assetCacheOut, amountOutDecoded);
+        cache.assetCacheOut.poolSize = decodeExternalAmount(cache.assetCacheOut, cache.preBalanceOut + amountOut);
+        AssetStorage storage assetStorageOut = eTokenLookup[cache.eTokenOut];
+        increaseBalance(assetStorageOut, cache.assetCacheOut, cache.eTokenOut, cache.accountOut, amountOutInternal);
+        logAssetStatus(cache.assetCacheOut);
+
+        // Check liquidity
+        checkLiquidity(cache.accountIn);
+        if (cache.accountIn != cache.accountOut && assetStorageOut.users[cache.accountOut].owed != 0)
+            checkLiquidity(cache.accountOut);
+    }
+
+    /// @notice Repay debt by selling another deposited token
+    /// @param subAccountIdIn sub-account holding the sold token. 0 for primary, 1-255 for a sub-account
+    /// @param subAccountIdOut sub-account to receive the bought token. 0 for primary, 1-255 for a sub-account
+    /// @param swapHandler address of a swap handler to use
+    /// @param params struct defining the requested trade
+    /// @param targetDebt how much debt should remain after calling the function
+    function swapAndRepay(uint subAccountIdIn, uint subAccountIdOut, address swapHandler, ISwapHandler.SwapParams memory params, uint targetDebt) external nonReentrant {
+        SwapCache memory cache = initSwap(subAccountIdIn, subAccountIdOut, params);
+
+        emit RequestSwapHubRepay(
+            cache.accountIn,
+            cache.accountOut,
+            params.underlyingIn,
+            params.underlyingOut,
+            targetDebt,
+            swapHandler
+        );
+
+        // Adjust params for repay
+        require(params.mode == 1, "e/swap-hub/repay-mode");
+        uint owed = getCurrentOwed(eTokenLookup[cache.eTokenOut], cache.assetCacheOut, cache.accountOut) / cache.assetCacheOut.underlyingDecimalsScaler;
+        require (owed > targetDebt, "e/swap-hub/target-debt");
+        params.amountOut = owed - targetDebt;
+
+        uint amountOut = swapInternal(cache, swapHandler, params);
+
+        // Process repay
+        AssetStorage storage assetStorageOut = eTokenLookup[cache.eTokenOut];
```

Comment:
may be moved above line 91 and used as `getCurrentOwed` input

Reply:
nice one!